sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/Scan.java

```diff
@@ -125,4 +125,26 @@ default CustomMetric[] supportedCustomMetrics() {
   default CustomTaskMetric[] reportDriverMetrics() {
     return new CustomTaskMetric[]{};
   }
+
+  /**
+   * This enum defines how the columnar support for the partitions of the data source
+   * should be determined. The default value is `PARTITION_DEFINED` which indicates that each
+   * partition can deterine if it should be columnar or not. SUPPORTED and UNSUPPORTED provide
+   * default shortcuts to indicate support for columnar data or not.
+   *
+   * @since 3.5.0
+   */
+  enum ColumnarSupportType {
+    PARTITION_DEFINED,
+    SUPPORTED,
+    UNSUPPORTED
+  }
+
+  /**
+   * Subclasses can implement this method to indicate if the support for columnar data should
+   * be determined by each partition or is set as a default for the whole scan.
+   *
+   * @since 3.5.0
+   */
+  default ColumnarSupportType supportsColumnar() { return ColumnarSupportType.PARTITION_DEFINED; }
```

Comment:
My initial impression is that the return value of this method should be Boolean type, as in Spark, other methods or functions with `support(s)` as a prefix are often like this.

Reply:
it can't be a boolean because there are three different values. 

Since this is used as an override, we must preserve today's behavior for backwards compatibility. The backward compatible behavior is to iterate over all partitions to check.

Reply:
then how about we name it `columnarSupportMode`?

Reply:
Will do!

Reply:
Renamed the method and enum type accordingly.

Reply:
Yeah, `columnarSupportMode ` if fine to me